README.md
# Layout grid helper

Направляющие верстки

![alt text](screenshots.png 'Пример работы плагина')

## Установка

```
```bash
# Using pnpm
pnpm add -D layout-grid-helper

# Using yarn
yarn add --dev layout-grid-helper

# Using npm
npm install --save-dev layout-grid-helper
```

## Использование

```js
import LayoutGridHelper from 'layout-grid-helper';

const gridHelper = new LayoutGridHelper({
	prefix: 'gh',
	sides: '16px',
	gutter: '16px',
	columns: 2,
	responsible: {
		640: {
			gutter: '20px',
			container: '620px',
			sides: '0px',
			columns: 4
		},
		960: {
			container: '940px',
			columns: 9
		},
		1600: {
			container: '1580px',
			columns: 10
		}
	}
});
```

## Опции

| Имя         | Описание                   | Значение по умолчанию                               |
|-------------|----------------------------|-----------------------------------------------------|
| className   | Класс блока сетки          | layout-grid-helper                                  |
| prefix      | Префикс переменных         | gh                                                  |
| mobileFirst | Приоритет мобильной версии | true                                                |
| color       | Цвет                       | rgb(255 0 0 / 0.2)                                  |
| gutter      | Отступ между колонками     | 16px                                                |
| sides       | Отступ от края экрана      | 20px                                                |
| columns     | Количество колонок         | 4                                                   |
| container   | Ширина контейнера          | 100%                                                |
| responsible | Медиа запросы              | {[breakpoint]:{gutter?,sides?,columns?,container?}} |

## Методы

| Имя     | Описание       |
|---------|----------------|
| show    | Показать сетку |
| hide    | Скрыть сетку   |
| destroy | Удаление сетки |

## Клавиатура

```
CTRL + M - покзать/скрыть сетку
```

## Благодарность

Библиотека сделана на основе [видео](https://youtu.be/WBrngvT78gw) Вадима Макеева


CHANGELOG.md
# Changelog


## v2.0.1

[compare changes](https://github.com/TakNePoidet/layout-grid-helper/compare/v2.0.0...v2.0.1)

### 📖 Documentation

- Fix documentation ([4a9972d](https://github.com/TakNePoidet/layout-grid-helper/commit/4a9972d))

### ❤️ Contributors

- Yakin Nikita ([@TakNePoidet](http://github.com/TakNePoidet))

## v2.0.0

[compare changes](https://github.com/TakNePoidet/layout-grid-helper/compare/v1.1.0...v2.0.0)

### 🚀 Enhancements

- Changing the batch manager ([421ec62](https://github.com/TakNePoidet/layout-grid-helper/commit/421ec62))
- Replacing rollup with vite ([766bd5f](https://github.com/TakNePoidet/layout-grid-helper/commit/766bd5f))
- Changing module type ([2722947](https://github.com/TakNePoidet/layout-grid-helper/commit/2722947))

### 💅 Refactors

- Rewriting in class ([ab664a3](https://github.com/TakNePoidet/layout-grid-helper/commit/ab664a3))

### 🎨 Styles

- Update styleguide ([6b35faa](https://github.com/TakNePoidet/layout-grid-helper/commit/6b35faa))
- Update eslint rules ([b5fb551](https://github.com/TakNePoidet/layout-grid-helper/commit/b5fb551))

### ❤️ Contributors

- Yakin Nikita ([@TakNePoidet](http://github.com/TakNePoidet))



rollup.config.ts
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import sourceMaps from 'rollup-plugin-sourcemaps';
import camelCase from 'lodash.camelcase';
import typescript from 'rollup-plugin-typescript2';
import json from 'rollup-plugin-json';

const pkg = require('./package.json');

const libraryName = 'layout-grid-helper';
const isDev = !process.env.NODE_ENV || process.env.NODE_ENV === 'development';

export default {
	input: `src/${libraryName}.ts`,
	output: [
		{ file: pkg.main, name: camelCase(libraryName), format: 'umd', sourcemap: isDev },
		{ file: pkg.module, format: 'es', sourcemap: isDev }
	],
	// Indicate here external modules you don't wanna include in your bundle (i.e.: 'lodash')
	external: [],
	watch: {
		include: 'src/**'
	},
	plugins: [
		json(),
		typescript({ useTsconfigDeclarationDir: true }),
		resolve(),
		commonjs(),
		sourceMaps()
	]
};

src/grid-hepler.ts

interface GridHelperVariables {
	gutter?: string;
	sides?: string;
	columns?: number;
	container?: string;
}

interface GridHelperConstructor extends GridHelperVariables {
	className?: string;
	prefix?: string;
	mobileFirst?: boolean,
	color?: string,
	responsible?: { [key: number]: GridHelperVariables; };
}
interface GridHelperApi {
	init(): GridHelperApi;
	show(): GridHelperApi;
	hide(): GridHelperApi;
	destroy(): void;
}



export default function GridHelper(params: GridHelperConstructor = {}): GridHelperApi {
	const color = params.color || 'rgb(255 0 0 / 0.2)';
	const gutter = params.gutter || '16px';
	const sides = params.sides || '20px';
	const columns = params.columns || 4;
	const container = params.container || '100%';
	const prefix = params.prefix || 'gh';
	const responsible = params.responsible || {};
	const className = params.className || 'grid-helper';
	const mobileFirst = typeof params.mobileFirst !== 'undefined' ? params.mobileFirst : true;
	let isShow = false;
	function createStyleElement(): HTMLStyleElement {
		const link = document.createElement('style');
		link.id = `id-grid-helper`;
		link.innerHTML = initStyleBase();
		return link;
	}
	function initStyleVariables(variables: GridHelperVariables): string {
		let style = ``;
		for (const key in variables) {
			if (Object.prototype.hasOwnProperty.call(variables, key)) {
				// @ts-ignore
				const value: string | number = variables[key];
				style += `--${prefix}-${key}: ${value};`;
			}
		}
		return style;
	}
	function initStyleBase(): string {
		const initVar = initStyleVariables({
			gutter,
			sides,
			columns,
			container,
		});
		let style = `
		.${className}::before {

			--${prefix}-color: ${color}; /* цвет колонки */
			${initVar}

			position: fixed;
			top: 0;
			width: 100%;
			height: 100%;
			content: '';
			pointer-events: none;
			z-index: 9999;
			background-repeat: no-repeat;
			max-width: var(--gh-container);
			left: 50%;
			transform: translateX(-50%);
			border: 0px solid transparent;

			--${prefix}-column-width: calc(
				((100% - (var(--${prefix}-gutter) * (var(--${prefix}-columns) - 1))) - var(--${prefix}-sides) * 2) / var(--${prefix}-columns)
			);
			
			margin: 0px var(--${prefix}-sides);
			/* 
				Градиент
			*/
			background-image: repeating-linear-gradient(
				90deg,
				var(--${prefix}-color) 0,
				var(--${prefix}-color) var(--${prefix}-column-width),
				transparent var(--${prefix}-column-width),
				transparent calc(var(--${prefix}-column-width) + var(--${prefix}-gutter))
			);
			
			
		}`;

		for (const size in responsible) {
			if (Object.prototype.hasOwnProperty.call(responsible, size)) {
				const vars = initStyleVariables(responsible[size]);
				style += `
					@media screen and (${mobileFirst ? 'min-width' : 'max-width'}: ${size}px) {
						.${className}::before {
							${vars}
						}
					}
				`;
			}
		}
		return style;
	}

	function keydown({ ctrlKey, code }: KeyboardEvent) {
		if (ctrlKey === true && code === "KeyM") {
			isShow ? hide() : show();
		}
	}
	function init() {
		document.head.append(createStyleElement());
		window.addEventListener('keydown', keydown);
		return api();
	}

	function destroy() {
		window.removeEventListener('keydown', keydown);
		const style = document.getElementById(`id-${className}`);
		if (style) {
			style.parentElement?.removeChild(style);
		}

	}
	function show() {
		isShow = true;
		document.body.classList.add(className);
		return api();
	}
	function hide() {
		isShow = false;
		document.body.classList.remove(className);
		return api();
	}
	function api(): GridHelperApi {
		return {
			init,
			show,
			hide,
			destroy
		};
	}
	return api();
}


src/layout-grid-helper.ts
import { defu } from 'defu';

interface Options {
	gutter: string;
	sides: string;
	columns: number;
	container: string;
}

export interface Params extends Options {
	className: string;
	prefix: string;
	color: string;
	mobileFirst: boolean;
	responsible: Record<number, Options>;
}

interface Api {
	active: boolean;
	show: () => void;
	hide: () => void;
	destroy: () => void;
}

export class LayoutGridHelper implements Api {
	readonly #options: Params;
	#show: boolean = false;
	#root = document.createElement('div');
	#style = document.createElement('style');
	#keydown: (event: KeyboardEvent) => void;

	constructor(params: Partial<Params> = {}) {
		this.#options = defu(params, {
			gutter: '16px',
			sides: '20px',
			columns: 4,
			container: '100%',
			className: 'layout-grid-helper',
			prefix: 'lgh',
			color: 'rgb(255 0 0 / 0.2)',
			mobileFirst: false,
			responsible: {}
		});

		this.#root.classList.add(this.#options.className);
		document.body.append(this.#root);
		this.#style.innerHTML = this.#render();
		document.head.append(this.#style);
		const keydown = ({ ctrlKey, code }: KeyboardEvent) => {
			if (ctrlKey && code === 'KeyM') {
				this.toggle();
			}
		};

		this.#keydown = keydown.bind(this);
		window.addEventListener('keydown', this.#keydown);
	}

	get active() {
		return this.#show;
	}

	set active(value) {
		this.#show = value;
		this.#root.style.display = value ? 'block' : 'none';
	}

	toggle(value?: boolean) {
		this.active = value ?? !this.active;
	}

	hide(): void {
		this.toggle(false);
	}

	show(): void {
		this.toggle(true);
	}

	destroy() {
		this.#root.remove();
		this.#style.remove();
		window.removeEventListener('keydown', this.#keydown);
	}

	#variable(key: string, value: string | undefined = undefined) {
		const { prefix } = this.#options;
		return `--${prefix}-${key}${value != null ? ` : ${value};` : ''}`;
	}

	#variables(variables: Partial<Options>): string {
		let style = '';

		for (const key in variables) {
			if (Object.prototype.hasOwnProperty.call(variables, key)) {
				const value = variables[key as keyof Options];

				if (typeof value === 'string' || typeof value === 'number') {
					style += this.#variable(key, value.toString());
				}
			}
		}
		return style;
	}

	#responsive() {
		const { responsible, mobileFirst, className } = this.#options;
		const keysResponsible = Object.keys(responsible)
			.map(Number)
			.sort((a: number, b: number) => {
				if (mobileFirst) {
					return a > b ? 1 : -1;
				}
				return a > b ? -1 : 1;
			});

		let style = '';

		for (const size of keysResponsible) {
			style += `
							@media (width ${mobileFirst ? '>' : '<'} ${size}px) {
								.${className} {
									${this.#variables(responsible[size])}
								}
							}
						`;
		}
		return style;
	}

	#render(): string {
		const { className, color, columns, container, gutter, sides } = this.#options;

		const variables = this.#variables({
			columns,
			container,
			gutter,
			sides
		});

		let columnWidthFormula = `((100% - (var(${this.#variable('gutter')})`;

		columnWidthFormula += ` * (var(${this.#variable('columns')}) - 1)))`;
		columnWidthFormula += ` + var(${this.#variable('sides')}) * 0) / var(${this.#variable('columns')})`;

		return `
				.${className} {
					${this.#variable('color', color)};
					${variables}
					${this.#variable('column-width', `calc(${columnWidthFormula})`)}
					display: none;
					position: fixed;
					top: 0;
					width: 100%;
					height: 100%;
					pointer-events: none;
					z-index: 9999;
					background-repeat: no-repeat;
					left: 50%;
					transform: translateX(-50%);
					border: 0px solid transparent;
					border-width: 0px  var(${this.#variable('sides')});
					max-width: var(${this.#variable('container')});
					background-image: repeating-linear-gradient(
						90deg,
						var(${this.#variable('color')}) 0,
						var(${this.#variable('color')}) var(${this.#variable('column-width')}),
						transparent var(${this.#variable('column-width')}),
						transparent calc(var(${this.#variable('column-width')}) + var(${this.#variable('gutter')}))
					);
				}
				
				${this.#responsive()}
			`;
	}
}
